Review messenger related documentation (#889)
api-platform/docs: core/messenger.md

/**
 * @ApiResource(
 *     collectionOperations={

>  *         "output"=false,

Reviewer: It will not work. Should be either as an attribute of `post` (preferred) or at the root of the annotation.

 *         "post"={"status"=202}
 *     },

>  *     messenger=true,

Reviewer: As only the `POST` operation is registered, wouldn't it be more explicit to set this attribute at the operation level too?

 * )
 */
final class Book
{
    /**
     * @var string
     *
     * @Assert\NotBlank
     */

>     public title;

Reviewer 2: ```suggestion
    public $title;
```

```yaml
# api/config/api_platform/resources.yaml
resources:
  App\Entity\Book:
    attributes:
        messenger: true
    collectionOperations:
      post:
        output: false
        status: 202
```

> Because the `messenger` attribute is `true`, when a write operation (`PUT`, `DELETE`, `POST`) is handled by API Platform, the corresponding instance of the `Book` will be dispatched.

Reviewer: ```suggestion
Because the `messenger` attribute is `true`, when a write operation (`PUT`, `DELETE`, `POST`) is handled by API Platform, after [deserialization](serialization.md) the hydrated instance of the `Book` object will be dispatched to [the default message bus](https://symfony.com/doc/current/messenger.html#dispatching-the-message).
```

> For the `POST` operation here, We use the `status` attribute to configure API Platform to return a [202 Accepted HTTP status code](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/202).

Reviewer: ```suggestion
For the `POST` operation here, we use the `status` attribute to return a [202 Accepted HTTP status code](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/202) in case of success.
```

By default, the handler will process your message synchronously.
If you want it to be consumed asynchronously (e.g. by a worker machine), [configure a transport and the consumer](https://symfony.com/doc/current/messenger.html#transports).

### Accessing the Data Returned by the Handler

API Platform automatically uses the `Symfony\Component\Messenger\Stamp\HandledStamp` when set.
It means that if you use a synchronous handler, the data returned by the `__invoke` method replaces the original data.

### Detecting Removals

When a `DELETE` operation occurs, API Platform automatically adds a `ApiPlatform\Core\Bridge\Symfony\Messenger\RemoveStamp` ["stamp"](https://symfony.com/doc/current/components/messenger.html#adding-metadata-to-messages-envelopes) instance to the "envelope".
To differentiate typical persists calls (create and update) and removal calls, check for the presence of this stamp using [a custom "middleware"](https://symfony.com/doc/current/components/messenger.html#adding-metadata-to-messages-envelopes).

> ### Working with Doctrine's entity manager

Reviewer: ```suggestion
### Working with Doctrine's Entity Manager
```

> Since Symfony 4.3, the messenger is using the native serializer (`Symfony\Component\Messenger\Transport\Serialization\PhpSerializer`). If you're using Doctrine entities as the Message, note that relation fields will not be in sync with Doctrine's Object Manager and you will have to merge those by hand in the message handler.

Reviewer: ```suggestion
When using async handlers, by default the Messenger component uses the native `serialize()` PHP function to serialize the content of the payload to transfer to the handler. If you're using a Doctrine entity as a message, it will not be in sync with the Doctrine's Object Manager when received by the handler. You'll have [to merge the entity](https://www.doctrine-project.org/projects/doctrine-orm/en/current/reference/working-with-objects.html#merging-entities) manually.
```

> Another solution is to switch the messenger's serializer to one supporting the format of your choice, for example:

Reviewer: I don't understand how serializing in JSON instead of using the native format fix prevent to have to merge the entity. I would drop the rest of this section entirely.

Reviewer 3: It's not about the format per se. It's because, by falling back to the default Symfony serializer, the Api-Platform item normalizer ends up being called and, since this object is a resource, that normalizer does the job of merging the entities back to the Object Manager. 

This was the default behavior before 4.3, which is why I suggested it should be mentioned somewhere.

Reviewer 3: I would leave the section and maybe add an extra caveat since it is entirely possible that a user reads the Symfony documentation and decides to use json for interoperability reasons and then don't realize why the behavior is different.

```
framework:
    messenger:
        serializer:
            default_serializer: messenger.transport.symfony_serializer
            symfony_serializer:
                format: json
                context: { }
```

> Note that when doing so it'll have a negative impact on performances. Also, if API Platform's serializer is used, it will work only for resources tagged as `ApiResource` and not for DTOs. Indeed, our normalizer expects resources to have an IRI, in other words at least one `GET` operation declared in symfony's `RouteCollection` (see also [](./operations.md)).

Reviewer 4: > Also, if API Platform's serializer is used, it will work only for resources tagged as `ApiResource` and not for DTOs.

I'm not sure I understand what you mean here.

Author: I'm trying to say that our normalizer doesn't support DTOs because they're not resources.

namespace App\Entity;

use ApiPlatform\Core\Annotation\ApiResource;
use App\Dto\PasswordResetRequest;

/**
 * @ApiResource(
 *     collectionOperations={
 *         "post",

>  *         "get",

Reviewer: should be first

 *         "post_password_reset_request"={
 *             "method"="POST",
 *             "path"="/users/password_reset_request",
 *             "messenger"="input",
 *             "input"=PasswordResetRequest::class,
 *             "status"=202,
 *             "output"=false,
 *             "normalization_context"={
 *                 "groups"={"user_password_reset_request"},
 *             },
 *             "denormalization_context"={
 *                 "groups"={"user_password_reset_request"},
 *             },
 *         },
 *     }
 * )
 */
final class User
{
}
```

Where `PasswordResetRequest` would be:

```php
<?php

// api/src/Dto/PasswordResetRequest.php

namespace App\Dto;

use Symfony\Component\Serializer\Annotation\Groups;

final class PasswordResetRequest
{
    /**
     * @Groups({"user_password_reset_request"})
     */

>     private $email;

Reviewer: Why not using a public properties? What's the benefit of using serialization groups here? It makes the config super verbose for no benefits.

    public function __construct(string $email = '')
    {
        $this->email = $email;
    }

    public function getEmail(): string
    {
        return $this->email;
    }
}
```

> For the `POST` operation here, We use the `status` attribute to configure API Platform to return a [202 Accepted HTTP status code](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/202).

Reviewer: duplicated, or keep "as above" maybe?
